examples/bri-3/src/bri/communication/agents/messaging.agent.ts

```diff
@@ -26,8 +29,92 @@ export class MessagingAgent implements OnApplicationBootstrap {
 
   private onNewMessageReceived(rawMessage: string): void {
     this.logger.logInfo(
-      `MessagingListenerAgent: New raw message received: ${rawMessage}. Dispatching ProcessInboundMessageCommand`,
+      `MessagingListenerAgent: New raw message received: ${rawMessage}. Trying to parse into a Bpi Message`,
+    );
+
+    const [newBpiMessageCandidate, validationErrors] =
+      this.validateBpiMessageFormat(rawMessage);
+
+    if (validationErrors.length > 0) {
+      this.logger.logError(
+        `MessagingListenerAgent: Failed parsing message into Bpi Message. Errors: ${validationErrors.join(
+          '; ',
+        )}`,
+      );
+
+      return;
+    }
+
+    this.commandBus.execute(
+      new ProcessInboundBpiMessageCommand(
+        newBpiMessageCandidate.id,
+        newBpiMessageCandidate.from,
+        newBpiMessageCandidate.to,
+        JSON.stringify(newBpiMessageCandidate.content),
+        newBpiMessageCandidate.signature,
+        newBpiMessageCandidate.type,
+      ),
+    );
+  }
+
+  public validateBpiMessageFormat(
+    rawMessage: string,
+  ): [CreateBpiMessageDto, string[]] {
+    const errors: string[] = [];
+    let newBpiMessageCandidate: CreateBpiMessageDto;
+
+    try {
+      newBpiMessageCandidate = this.parseJsonOrThrow(rawMessage);
+    } catch (e) {
+      errors.push(`${rawMessage} is not valid JSON. Error: ${e}`);
+      return [newBpiMessageCandidate, errors];
+    }
+
+    if (!validate(newBpiMessageCandidate.id)) {
+      errors.push(`id ${newBpiMessageCandidate.id} is not a valid UUID`);
+    }
+
+    if (!validate(newBpiMessageCandidate.from)) {
```

Comment:
Can you explain the logical division between the bpiMessages.agent and the messaging.agent?

Reply:
I also thought about better naming but failed to produce anything meaningful. Open for proposals as we go.

bpiMessages agent deals with BpiMessages, which is the internal domain of the BPI.

messagging.agent lives on the boundary of the internal domain and the outside world and deals with translating BpiMessages to and from message formats used by various messaging clients (atm only NATS in plain text).